plugins/ata/fu-ata-device.c

```diff
@@ -0,0 +1,603 @@
+		g_set_error_literal (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "DOWNLOAD_MICROCODE not supported by device");
+		return FALSE;
+	}
+
+	/* fallback to a sane default */
+	if (self->transfer_mode == 0) {
+		if ((id[119] & 0x10) && (id[120] & 0x10))
+			self->transfer_mode = 0x3;
+		else
+			self->transfer_mode = 0x7;
+	}
+
+	/* the newer, segmented transfer mode */
+	if (self->transfer_mode == 0x3 || self->transfer_mode == 0xe) {
+		xfer_min = id[234];
+		if (xfer_min == 0x0 || xfer_min == 0xffff)
+			xfer_min = 1;
+		xfer_max = id[235];
+		if (xfer_max == 0x0 || xfer_max == 0xffff)
+			xfer_max = xfer_min;
+	}
+
+	/* fall back to a sane block size */
+	if (self->transfer_size == 0x0)
+		self->transfer_size = xfer_min * FU_ATA_BLOCK_SIZE;
+	else if (self->transfer_size == 0xffff)
+		self->transfer_size = xfer_max * FU_ATA_BLOCK_SIZE;
+
+	/* get values in case the kernel didn't */
+	if (fu_device_get_serial (device) == NULL) {
+		g_autofree gchar *tmp = NULL;
+		tmp = fu_ata_device_get_string (id, 10, 19);
+		fu_device_set_serial (device, tmp);
+	}
+	if (fu_device_get_name (device) == NULL) {
+		g_autofree gchar *tmp = NULL;
+		tmp = fu_ata_device_get_string (id, 27, 46);
+		fu_device_set_name (device, tmp);
+	}
+	if (fu_device_get_version (device) == NULL) {
+		g_autofree gchar *tmp = NULL;
+		tmp = fu_ata_device_get_string (id, 23, 26);
+		fu_device_set_version (device, tmp);
+	}
+
+	/* 8 byte additional product identifier == SKU? */
+	sku = fu_ata_device_get_string (id, 170, 173);
+	if (sku != NULL)
+		g_debug ("SKU=%s", sku);
+
+	/* add extra GUIDs */
+	name_pad = fu_ata_device_pad_string_for_id (fu_device_get_name (device));
+	if (name_pad != NULL &&
+	    fu_device_get_version (device) != NULL) {
+		g_autofree gchar *tmp = NULL;
+		tmp = g_strdup_printf ("IDE\\%s%s", name_pad,
+				       fu_device_get_version (device));
+		fu_device_add_guid (device, tmp);
+	}
+	if (name_pad != NULL) {
+		g_autofree gchar *tmp = NULL;
+		tmp = g_strdup_printf ("IDE\\0%s", name_pad);
+		fu_device_add_guid (device, tmp);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_ata_device_open (FuDevice *device, GError **error)
+{
+	FuAtaDevice *self = FU_ATA_DEVICE (device);
+	GUdevDevice *udev_device = fu_udev_device_get_dev (FU_UDEV_DEVICE (device));
+
+	/* open device */
+	self->fd = g_open (g_udev_device_get_device_file (udev_device), O_RDONLY);
+	if (self->fd < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_FAILED,
+			     "failed to open %s: %s",
+			     g_udev_device_get_device_file (udev_device),
+			     strerror (errno));
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_ata_device_probe (FuUdevDevice *device, GError **error)
+{
+	FuAtaDevice *self = FU_ATA_DEVICE (device);
+
+	/* set the physical ID */
+	if (!fu_udev_device_set_physical_id (device, "scsi", error))
+		return FALSE;
+
+	/* look at the PCI depth to work out if in an external enclosure */
```

Comment:
PCI?  Shouldn't ATA controllers typically be connected off the PCH?  Or is this for betting against an AHCI controller on a PCIe card inserted into a thunderbolt enclosure?

Reply:
Yes, I'm trying to guard against the TBT enclosure thing.

Reply:
OK